### PR TITLE
CoreCLR NativeAOT - ByRef generics and ByRef implementing interfaces

### LINKED ISSUE #68002: [API Proposal]: Add managed API surface area for ByRefLike Generic constraints
### Background and motivation

The [design for ByRefLike Generic constraints](https://github.com/dotnet/runtime/pull/67129) requires some new API surface area. Specifically the need to represent the new metadata values and a new feature flag to indicate support for the compiler.

### API Proposal

Metadata enum values:
```diff
namespace System.Reflection
{
    [Flags]
    public enum GenericParameterAttributes
    {
+        [Obsolete("No longer represents all special constraints")]
         SpecialConstraintMask = 0x001C,
+        AcceptByRefLike = 0x0020,
    }
}
```

Feature flag.
```diff
namespace System.Runtime.CompilerServices
{
    public static partial class RuntimeFeature
    {
+        /// <summary>
+        /// Represents a runtime feature where byref-like types can be used in Generic parameters.
+        /// </summary>
+        public const string GenericsAcceptByRefLike = nameof(GenericsAcceptByRefLike);
    }
}
```

### API Usage

Usage is similar to the official documentation for any of the existing [`GenericParameterAttributes`](https://docs.microsoft.com/dotnet/api/system.reflection.genericparameterattributes) enumeration values.

### Alternative Designs

The following would represent a binary breaking change.

```diff
namespace System.Reflection
{
    [Flags]
    public enum GenericParameterAttributes
    {
-        SpecialConstraintMask = 0x001C,
+        SpecialConstraintMask = 0x003C,
+        AcceptByRefLike = 0x0020,
    }
}
```

### Risks

Updating the enumeration value is troublesome since it contains a value that is a mask&mdash;an established anti-pattern. The "safest" way to handle this is the initial proposal but that does introduce confusion about new scenarios. It might be prudent to create a new API for accessing the specific constraints from the variance values.

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Contributes to https://github.com/dotnet/runtime/issues/65112
Fixes https://github.com/dotnet/runtime/issues/68002

### CODE REVIEW

File: src/coreclr/tools/Common/Compiler/CompilerTypeSystemContext.Validation.cs [hunk 1]
```diff
@@ -368,6 +368,16 @@ private static TypeDesc EnsureLoadableTypeUncached(TypeDesc type)
                     ThrowHelper.ThrowTypeLoadException(ExceptionStringID.ClassLoadGeneral, type);
                 }
 
+                // Validate fields meet the constraints of the enclosing type.
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/Compiler/CompilerTypeSystemContext.Validation.cs:371]: I think this is the kind of check we could leave out from AOT compilers. A field not meeting constraints doesn't seem any more damaging than a local or parameter and we don't check those and probably wouldn't want to spend time checking them.

Comment (AaronRobinsonMSFT) [reply src/coreclr/tools/Common/Compiler/CompilerTypeSystemContext.Validation.cs:371]: Okay.

**Comment by AaronRobinsonMSFT** [comment]
FYI @fanyang-mono @lambdageek

File: src/coreclr/nativeaot/System.Private.Reflection.Execution/src/Internal/Reflection/Execution/TypeLoader/ConstraintValidator.cs [hunk 2]
```diff
@@ -13,7 +13,7 @@ internal static partial class ConstraintValidator
     {
         private static bool SatisfiesConstraints(this Type genericVariable, SigTypeContext typeContextOfConstraintDeclarer, Type typeArg)
         {
-            GenericParameterAttributes specialConstraints = genericVariable.GenericParameterAttributes & GenericParameterAttributes.SpecialConstraintMask;
+            GenericParameterAttributes specialConstraints = genericVariable.GenericParameterAttributes;
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/System.Private.Reflection.Execution/src/Internal/Reflection/Execution/TypeLoader/ConstraintValidator.cs:16]: Nit: This is no longer just the special constrains.

```suggestion
            GenericParameterAttributes attributes = genericVariable.GenericParameterAttributes;
```

**Review by lambdageek — APPROVED** [review]

**Pull request merged**